rllib/models/specs/specs_base.py

```diff
@@ -0,0 +1,209 @@
+        if len(shape) != len(self._expected_shape):
+            raise ValueError(_INVALID_SHAPE.format(self._expected_shape, shape))
+
+        for expected_d, actual_d in zip(self._expected_shape, shape):
+            if isinstance(expected_d, int) and expected_d != actual_d:
+                raise ValueError(_INVALID_SHAPE.format(self._expected_shape, shape))
+
+        dtype = self.get_dtype(tensor)
+        if self.dtype and dtype != self.dtype:
+            raise ValueError(_INVALID_DTYPE.format(self.dtype, tensor.dtype))
+
+    @abc.abstractmethod
+    def get_shape(self, tensor: TensorType) -> Tuple[int]:
+        """Returns the shape of a tensor.
+
+        Args:
+            tensor: The tensor whose shape is to be returned.
+
+        Returns:
+            A `tuple` specifying the shape of the tensor.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def get_dtype(self, tensor: TensorType) -> Any:
+        """Returns the data type of a tensor.
+
+        Args:
+            tensor: The tensor whose data type is to be returned.
+
+        Returns:
+            The data type of the tensor.
+        """
+        raise NotImplementedError
+
+    @DeveloperAPI
+    def sample(self, fill_value: Union[float, int] = 0) -> TensorType:
+        """Samples a tensor with the specified value that is matches the specs.
+
+        Args:
+            fill_value: The value to fill the tensor with.
+
+        Returns:
+            A tensor with the specified value that matches the specs.
+        """
+        sampled_shape = self._sample_shape()
+        return self._sample(sampled_shape, fill_value)
+
+    @abc.abstractmethod
+    def _sample(
+        self, shape: Tuple[int], fill_value: Union[float, int] = 0
+    ) -> TensorType:
+        """Returns a sample tensor with the given shape.
+
+        Args:
+            shape: The shape of the tensor to be sampled.
+            fill_value: The value to fill the tensor with.
+
+        Returns:
+            A tensor with the specified value that matches the specs.
+        """
+        raise NotImplementedError
+
+    def _sample_shape(self) -> Tuple[int]:
+        """Converts the expected shape to a sample shape by replacing TBD dimensions
+        with size of 1."""
+        sampled_shape = tuple()
+        for d in self._expected_shape:
+            if isinstance(d, int):
+                sampled_shape += (d,)
+            else:
+                sampled_shape += (1,)
+        return sampled_shape
+
+    def _parse_expected_shape(self, shape: str, shape_vals: Dict[str, int]) -> tuple:
+        """Converts the input shape to a tuple of integers and strings."""
+        expected_shape = tuple()
+
+        # check the validity of shape_vals and get a list of dimension names
+        d_names = self._validate_shape_vals(shape, shape_vals)
+
+        for d in d_names:
+            d_value = shape_vals.get(d, None)
+            if d_value is None:
```

Comment:
Hmm I have a question. Is batch the only variable length dimension that we'll have to ever deal with? and if so can we just make sure that all of the other non batch dimensions have shape included?

Reply:
this is probably my biggest question.

Reply:
No it's not just batch. It could be time in RNNs so `("B T h", h=self.hidden_dim)`, in CNNs it could be `("B C H W", H=self.size, W=self.size)` and potentially anything else that is not invented yet :)

Reply:
Graphs could be `batch graphs nodes` which could all be variable-sized

Reply:
oh actually this is an important use case. should support size=None maybe for some dimensions that may be variable length?

Reply:
if you don't specify it, it will be interpreted as variable dimension, like B for example.